<html>
  <head>
    <!-- Google tag (gtag.js) -->
    <script
      async
      src="https://www.googletagmanager.com/gtag/js?id=G-6P1ZQ0CC8F"
    ></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag() {
        dataLayer.push(arguments);
      }
      gtag("js", new Date());

      gtag("config", "G-6P1ZQ0CC8F");
    </script>
    <meta content="text/html; charset=UTF-8" http-equiv="content-type" />
    <style type="text/css">
      ol {
        margin: 0;
        padding: 0;
      }
      table td,
      table th {
        padding: 0;
      }
      .c10 {
        border-right-style: solid;
        padding: 5pt 5pt 5pt 5pt;
        border-bottom-color: #808080;
        border-top-width: 0pt;
        border-right-width: 0pt;
        border-left-color: #808080;
        vertical-align: top;
        border-right-color: #808080;
        border-left-width: 0pt;
        border-top-style: solid;
        border-left-style: solid;
        border-bottom-width: 0pt;
        width: 126.2pt;
        border-top-color: #808080;
        border-bottom-style: solid;
      }
      .c5 {
        border-right-style: solid;
        padding: 5pt 5pt 5pt 5pt;
        border-bottom-color: #808080;
        border-top-width: 0pt;
        border-right-width: 0pt;
        border-left-color: #808080;
        vertical-align: top;
        border-right-color: #808080;
        border-left-width: 0pt;
        border-top-style: solid;
        border-left-style: solid;
        border-bottom-width: 0pt;
        width: 100pt;
        border-top-color: #808080;
        border-bottom-style: solid;
      }
      .c7 {
        border-right-style: solid;
        padding: 5pt 5pt 5pt 5pt;
        border-bottom-color: #808080;
        border-top-width: 0pt;
        border-right-width: 0pt;
        border-left-color: #808080;
        vertical-align: top;
        border-right-color: #808080;
        border-left-width: 0pt;
        border-top-style: solid;
        border-left-style: solid;
        border-bottom-width: 0pt;
        width: 94.8pt;
        border-top-color: #808080;
        border-bottom-style: solid;
      }
      .c0 {
        margin-left: 30pt;
        padding-top: 12pt;
        padding-bottom: 12pt;
        line-height: 1.15;
        orphans: 2;
        widows: 2;
        text-align: left;
        margin-right: 30pt;
      }
      .c1 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-size: 11pt;
        font-family: "Arial";
        font-style: normal;
      }
      .c14 {
        color: #000000;
        text-decoration: none;
        vertical-align: baseline;
        font-size: 17pt;
        font-family: "Arial";
        font-style: normal;
      }
      .c12 {
        color: #000000;
        text-decoration: none;
        vertical-align: baseline;
        font-size: 11pt;
        font-family: "Arial";
        font-style: normal;
      }
      .c9 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-size: 11pt;
        font-family: "Arial";
      }
      .c21 {
        font-weight: 400;
        vertical-align: baseline;
        font-size: 11pt;
        font-family: "Arial";
        font-style: normal;
      }
      .c17 {
        padding-top: 0pt;
        padding-bottom: 0pt;
        line-height: 1.15;
        text-align: center;
      }
      .c6 {
        padding-top: 12pt;
        padding-bottom: 12pt;
        line-height: 1.15;
        text-align: left;
      }
      .c3 {
        padding-top: 0pt;
        padding-bottom: 0pt;
        line-height: 1.15;
        text-align: left;
      }
      .c19 {
        padding-top: 18pt;
        padding-bottom: 4pt;
        line-height: 1.15;
        text-align: left;
      }
      .c20 {
        border-spacing: 0;
        border-collapse: collapse;
        margin-right: auto;
      }
      .c16 {
        text-decoration-skip-ink: none;
        -webkit-text-decoration-skip: none;
        color: #1155cc;
        text-decoration: underline;
      }
      .c18 {
        background-color: #ffffff;
        max-width: 468pt;
        padding: 72pt 72pt 72pt 72pt;
      }
      .c2 {
        color: inherit;
        text-decoration: inherit;
      }
      .c4 {
        orphans: 2;
        widows: 2;
      }
      .c13 {
        height: 25pt;
      }
      .c11 {
        font-weight: 700;
      }
      .c8 {
        font-style: italic;
      }
      .c15 {
        height: 11pt;
      }
      .title {
        padding-top: 0pt;
        color: #000000;
        font-size: 26pt;
        padding-bottom: 3pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      .subtitle {
        padding-top: 0pt;
        color: #666666;
        font-size: 15pt;
        padding-bottom: 16pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      li {
        color: #000000;
        font-size: 11pt;
        font-family: "Arial";
      }
      p {
        margin: 0;
        color: #000000;
        font-size: 11pt;
        font-family: "Arial";
      }
      h1 {
        padding-top: 20pt;
        color: #000000;
        font-size: 20pt;
        padding-bottom: 6pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      h2 {
        padding-top: 18pt;
        color: #000000;
        font-size: 16pt;
        padding-bottom: 6pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      h3 {
        padding-top: 16pt;
        color: #434343;
        font-size: 14pt;
        padding-bottom: 4pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      h4 {
        padding-top: 14pt;
        color: #666666;
        font-size: 12pt;
        padding-bottom: 4pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      h5 {
        padding-top: 12pt;
        color: #666666;
        font-size: 11pt;
        padding-bottom: 4pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      h6 {
        padding-top: 12pt;
        color: #666666;
        font-size: 11pt;
        padding-bottom: 4pt;
        font-family: "Arial";
        line-height: 1.15;
        page-break-after: avoid;
        font-style: italic;
        orphans: 2;
        widows: 2;
        text-align: left;
      }
      body {
        width: 50%;
        margin: 0 auto;
      }

      blockquote {
        margin: 0;
      }

      blockquote p {
        padding: 15px;
        background: #eee;
        border-radius: 5px;
      }

      a,
      a:hover,
      a:visited,
      a:active {
        color: inherit;
      }
      a:hover {
        border-bottom: 1px dotted red;
        color: red;
      }
    </style>
  </head>
  <body>
    <p style="color: inherit; text-decoration: none">
      <a href="../index.html">Home</a>
      <a href="../blog.html">Blog</a>
    </p>
    <h1>FASTerQ to Count Matrices for Single-Cell RNA-Seq</h1>
    <p class="c4 c6">
      <span class="c12 c11"
        >tl;dr Stream reads from the ENA to generate count matrices without ever
        saving the FASTQs using kallisto | bustools; faster than downloading the
        FASTQs.</span
      >
    </p>
    <hr />
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c6 c4">
      <span>The UNIX operating system was originally</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.britannica.com/technology/UNIX&amp;sa=D&amp;source=editors&amp;ust=1671999727973874&amp;usg=AOvVaw0Vkhf7W05nrq5uswT2iral"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.britannica.com/technology/UNIX&amp;sa=D&amp;source=editors&amp;ust=1671999727974201&amp;usg=AOvVaw1BMGMsj3bL9gbrnOY6WDRo"
          >developed at Bell Labs</a
        ></span
      ><span class="c1"
        >&nbsp;for research purposes in the 1960s and has developed today into
        operating systems such as Mac OS and GNU/Linux. One of the
        &ldquo;maxims&rdquo; that took hold among the initial builders and users
        of UNIX systems was the following:</span
      >
    </p>
    <p class="c0">
      <span class="c9 c8"
        >Expect the output of every program to become the input to another, as
        yet unknown, program. Don&#39;t clutter output with extraneous
        information. Avoid stringently columnar or binary input formats.
        Don&#39;t insist on interactive input.</span
      >
    </p>
    <p class="c0">
      <span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://archive.org/details/bstj57-6-1899/page/n3&amp;sa=D&amp;source=editors&amp;ust=1671999727974852&amp;usg=AOvVaw0_DVuXjJ0LrwFBxXWWjEMG"
          >UNIX Time-Sharing System: Forward. McIlroy, M.D.; Pinson, E.N.;
          Tague, B.A.</a
        ></span
      >
    </p>
    <p class="c6 c4">
      <span>This maxim is </span><span class="c11">extremely</span
      ><span
        >&nbsp;powerful and thanks to P&aacute;ll Melsted&rsquo;s foresight and
        execution, was adhered to during the development of </span
      ><span class="c11">kallisto</span><span>&nbsp;(</span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.nature.com/articles/nbt.3519&amp;sa=D&amp;source=editors&amp;ust=1671999727975434&amp;usg=AOvVaw0a_ckXHk5uq4USBPzEQ6ms"
          >Bray et al.</a
        ></span
      ><span>) and now the </span><span class="c11">kallisto | bustools </span
      ><span>single-cell pre-processing workflow (</span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.biorxiv.org/content/10.1101/673285v1&amp;sa=D&amp;source=editors&amp;ust=1671999727975812&amp;usg=AOvVaw2beoM-IO8dtJBMK1Gq_9Kd"
          >Melsted, Booeshaghi et al.</a
        ></span
      ><span class="c1">).</span>
    </p>
    <p class="c6 c4">
      <span
        >Pre-processing single-cell RNA-seq starts with downloading large FASTQ
        files (Figure 1) and then aligning the reads using your favorite
        alignment program. Anyone who has lived through the</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.buzzfeed.com/mjs538/please-be-done-downloading&amp;sa=D&amp;source=editors&amp;ust=1671999727976188&amp;usg=AOvVaw34nb-fHRBV6sx4P7qPa07D"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.buzzfeed.com/mjs538/please-be-done-downloading&amp;sa=D&amp;source=editors&amp;ust=1671999727976408&amp;usg=AOvVaw0yJzn_xCTK6YD9PFDDaq1E"
          >age of Limewire</a
        ></span
      ><span class="c1"
        >&nbsp;knows that downloading, while much faster nowadays, is still
        incredibly slow for gigabyte sized FASTQ files.
      </span>
    </p>
    <p class="c3 c4">
      <span
        style="
          overflow: hidden;
          display: inline-block;
          margin: 0px 0px;
          border: 0px solid #000000;
          transform: rotate(0rad) translateZ(0px);
          -webkit-transform: rotate(0rad) translateZ(0px);
          width: 624px;
          height: 602.67px;
        "
        ><img
          alt=""
          src="img_fasterq/image2.png"
          style="
            width: 624px;
            height: 602.67px;
            margin-left: 0px;
            margin-top: 0px;
            transform: rotate(0rad) translateZ(0px);
            -webkit-transform: rotate(0rad) translateZ(0px);
          "
          title=""
      /></span>
    </p>
    <p class="c3 c4">
      <span class="c8 c9"
        >Figure 1: The size of the kallisto | bustools files (y-axis) vs. the
        size of the FASTQ files (x-axis) for 21 single-cell RNA-Seq
        datasets.</span
      >
    </p>
    <p class="c6 c4">
      <span
        >Frustrated by having to wait for my reads to download before I could
        process them and inspired by the design of bustools </span
      ><span class="c11"
        >I sought a way to download and stream reads into the kallisto |
        bustools workflow from a download link</span
      ><span class="c1"
        >. This would allow me to (a) never have to write the FASTQ files to
        disk, thereby conserving disk space and (b) obtain gene count matrices
        in the time it takes to download FASTQ files.
      </span>
    </p>
    <p class="c6 c4">
      <span>After a bit of Google-fu I came across a UNIX object called a</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=http://man7.org/linux/man-pages/man7/fifo.7.html&amp;sa=D&amp;source=editors&amp;ust=1671999727977215&amp;usg=AOvVaw0rkMLNpMczoSdwzTvUYPqb"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=http://man7.org/linux/man-pages/man7/fifo.7.html&amp;sa=D&amp;source=editors&amp;ust=1671999727977490&amp;usg=AOvVaw3MzWasi68sbeRZC5wON-l7"
          >named-pipe</a
        ></span
      ><span
        >. A named-pipe is essentially a temporary &ldquo;file&rdquo; that data
        can be written to and read from in a FIFO (first in first out) manner. A
        neat quirk of named-pipes is that data that is read from it is removed
        from it. So if we read data from the named-pipe at the same speed that
        we write to it (or faster), the disk footprint of the named pipe is
        effectively zero. Making named-pipes is easy with the</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/3/mkfifo&amp;sa=D&amp;source=editors&amp;ust=1671999727977777&amp;usg=AOvVaw3U_TZGCrv-IYY8rAs02TNj"
          >&nbsp;</a
        ></span
      ><span class="c16 c11"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/3/mkfifo&amp;sa=D&amp;source=editors&amp;ust=1671999727977978&amp;usg=AOvVaw2OqD2UMKPUr7j2AANpEoey"
          >mkfifo</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/3/mkfifo&amp;sa=D&amp;source=editors&amp;ust=1671999727978166&amp;usg=AOvVaw3AdTgzG7nJ29tNpWyZ9eXA"
          >&nbsp;command</a
        ></span
      ><span class="c1">.</span>
    </p>
    <p class="c6 c4">
      <span
        >Now all that we have to do is figure out a way to download data and
        stream them to the named pipe. That is where </span
      ><span class="c11">curl </span><span>comes in to play. The</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/1/curl&amp;sa=D&amp;source=editors&amp;ust=1671999727978581&amp;usg=AOvVaw2Ih1tbULnSp2R_i7hKi9ih"
          >&nbsp;</a
        ></span
      ><span class="c16 c11"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/1/curl&amp;sa=D&amp;source=editors&amp;ust=1671999727978832&amp;usg=AOvVaw3P7rBRx1X0CHxnt2HvKbTM"
          >curl</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://linux.die.net/man/1/curl&amp;sa=D&amp;source=editors&amp;ust=1671999727979042&amp;usg=AOvVaw1qyntCfdljW5mE2O1ClrZm"
          >&nbsp;command</a
        ></span
      ><span
        >&nbsp;allows users to download files from a link and in the typical
        UNIX fashion </span
      ><span class="c11">curl</span
      ><span class="c1"
        >&nbsp;allows one to stream the data to a file as it is being
        downloaded.</span
      >
    </p>
    <h2 class="c4 c19" id="h.u5ckcyt41nvp">
      <span class="c14 c11">A worked example</span>
    </h2>
    <p class="c6 c4">
      <span>Using </span><span class="c11">curl</span><span>&nbsp;and </span
      ><span class="c11">mkfifo</span
      ><span>&nbsp;we can stream reads from a download link directly to </span
      ><span class="c11">kallisto bus</span
      ><span>&nbsp;and subsequently to </span><span class="c11">bustools</span
      ><span
        >. Here is an example you can try for yourself on your UNIX-based
        terminal. I am following the </span
      ><span class="c11">kallisto | bustools</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.kallistobus.tools/getting_started&amp;sa=D&amp;source=editors&amp;ust=1671999727980163&amp;usg=AOvVaw2RJZ3usWdz7nprKT_980Ml"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.kallistobus.tools/getting_started&amp;sa=D&amp;source=editors&amp;ust=1671999727980441&amp;usg=AOvVaw1CQMJlxtKltk_lILIbAHiZ"
          >Getting Started tutorial</a
        ></span
      ><span>&nbsp;processing 8,860,361 reads from single</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.ncbi.nlm.nih.gov/sra/?term%3DSRR8599150&amp;sa=D&amp;source=editors&amp;ust=1671999727980752&amp;usg=AOvVaw0FCyfDFy850OxwM3eMtTmK"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.ncbi.nlm.nih.gov/sra/?term%3DSRR8599150&amp;sa=D&amp;source=editors&amp;ust=1671999727980972&amp;usg=AOvVaw3_vYfNqXX1MNIwxGSURmG2"
          >mouse retinal cells SRR8599150</a
        ></span
      ><span>&nbsp;(</span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://doi.org/10.1016/j.immuni.2019.02.007&amp;sa=D&amp;source=editors&amp;ust=1671999727981192&amp;usg=AOvVaw0biNIWkhW9w9EC1OmEZQx4"
          >Koren et al., 2019</a
        ></span
      ><span class="c1">).</span>
    </p>
    <p class="c6 c4">
      <span>Once you have installed the necessary programs (</span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.kallistobus.tools/downloads&amp;sa=D&amp;source=editors&amp;ust=1671999727981524&amp;usg=AOvVaw28H6RpT3bKipaCTEBrs1m5"
          >kallisto and bustools</a
        ></span
      ><span>) then you can proceed. First install the necessary files (</span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/BUStools/getting_started/releases/tag/getting_started&amp;sa=D&amp;source=editors&amp;ust=1671999727981863&amp;usg=AOvVaw0kFlG0w91_A36jhLw9HmVg"
          >index, whitelist, and transcripts_to_genes</a
        ></span
      ><span class="c1">).</span>
    </p>
    <p class="c3 c4"><span class="c1">## Download necessary files</span></p>
    <p class="c3 c4">
      <span class="c1"
        >$ curl -L
        https://github.com/BUStools/getting_started/releases/download/getting_started/Mus_musculus.GRCm38.cdna.all.idx.gz
        -o Mus_musculus.GRCm38.cdna.all.idx.gz</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1"
        >$ curl -L
        https://github.com/BUStools/getting_started/releases/download/getting_started/10xv2_whitelist.txt
        -o 10xv2_whitelist.txt</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1"
        >$ curl -L
        https://github.com/BUStools/getting_started/releases/download/getting_started/transcripts_to_genes.txt
        -o transcripts_to_genes.txt</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1">$ gunzip Mus_musculus.GRCm38.cdna.all.idx.gz</span>
    </p>
    <p class="c6 c4">
      <span class="c1"
        >Now you can stream reads. The code needed to do so is a single line
        which you can past onto the command line. Below we provide both the
        line, and the &quot;expanded&quot; line for clarity.</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1">## One-liner: Stream to Count Matrices</span>
    </p>
    <p class="c3 c4">
      <span class="c1"
        >$ mkfifo R1.gz R2.gz; curl -Ls
        https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R1_001.fastq.gz
        &nbsp;&gt; R1.gz &amp; curl -Ls
        https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R2_001.fastq.gz
        &nbsp;&gt; R2.gz &amp; kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx
        -x 10xv2 -t 4 -o bus_out/ R1.gz R2.gz; cd bus_out/; mkdir tmp/; bustools
        correct -w ../10xv2_whitelist.txt -p output.bus | bustools sort -t 4 -T
        tmp/ -p - | bustools count -o genes -g ../transcripts_to_genes.txt -e
        matrix.ec -t transcripts.txt --genecounts -</span
      >
    </p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c3 c4">
      <span class="c1">## Expanded: Stream to Count Matrices</span>
    </p>
    <p class="c3 c4"><span class="c1">$ mkfifo R1.gz R2.gz</span></p>
    <p class="c3 c4">
      <span class="c1"
        >$ curl -Ls
        https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R1_001.fastq.gz
        &nbsp;&gt; R1.gz &amp;</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1"
        >$ curl -Ls
        https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R2_001.fastq.gz
        &nbsp;&gt; R2.gz &amp;</span
      >
    </p>
    <p class="c3 c4">
      <span class="c1"
        >$ kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx -x 10xv2 -t 4 -o
        bus_out/ R1.gz R2.gz</span
      >
    </p>
    <p class="c3 c4"><span class="c1">$ cd bus_out/</span></p>
    <p class="c3 c4"><span class="c1">$ mkdir tmp/</span></p>
    <p class="c3 c4">
      <span class="c1"
        >$ bustools correct -w ../10xv2_whitelist.txt -p output.bus | bustools
        sort -t 4 -T tmp/ -p - | bustools count -o genes -g
        ../transcripts_to_genes.txt -e matrix.ec -t transcripts.txt --genecounts
        -</span
      >
    </p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c6 c4">
      <span>I went ahead and made</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/sbooeshaghi/web_log/tree/master/2019_07_07-FASTerQ_to_Count_Matrices_for_Single_Cell_RNA_Seq&amp;sa=D&amp;source=editors&amp;ust=1671999727984316&amp;usg=AOvVaw1X9KX75BGgUiKdMHeT3PDD"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/sbooeshaghi/web_log/tree/master/2019_07_07-FASTerQ_to_Count_Matrices_for_Single_Cell_RNA_Seq&amp;sa=D&amp;source=editors&amp;ust=1671999727984686&amp;usg=AOvVaw1pTjNk2p8P_xN6DPDLdRew"
          >two Snakemakes</a
        ></span
      ><span
        >&nbsp;to compare the runtimes of (a) stream processing: streaming the
        reads as explained above and (b) download processing: downloading FASTQs </span
      ><span class="c8">then</span><span>&nbsp;pre-processing. </span
      ><span class="c11"
        >The streaming workflow processes data faster than it takes to download
        the data.</span
      ><span class="c1">&nbsp;Here are the raw numbers:</span>
    </p>
    <p class="c3 c4">
      <span class="c1">DATASET: SRR8599150 (8,860,361 reads)</span>
    </p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c3 c4"><span class="c1">STREAM + RUN</span></p>
    <p class="c3 c4"><span class="c1">download time: 0 sec</span></p>
    <p class="c3 c4"><span class="c1">kallisto time: 93 sec</span></p>
    <p class="c3 c4"><span class="c1">bustools time: 17 sec</span></p>
    <p class="c3 c4"><span class="c1">Total: 110 sec</span></p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c3 c4"><span class="c1">DOWNLOAD + RUN</span></p>
    <p class="c3 c4"><span class="c1">download time: 112 sec</span></p>
    <p class="c3 c4"><span class="c1">kallisto time: 41 sec</span></p>
    <p class="c3 c4"><span class="c1">bustools time: 17 sec</span></p>
    <p class="c3 c4"><span class="c1">Total: 170 sec</span></p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
    <p class="c17 c4">
      <span
        style="
          overflow: hidden;
          display: inline-block;
          margin: 0px 0px;
          border: 0px solid #000000;
          transform: rotate(0rad) translateZ(0px);
          -webkit-transform: rotate(0rad) translateZ(0px);
          width: 451.5px;
          height: 301px;
        "
        ><img
          alt=""
          src="img_fasterq/image1.png"
          style="
            width: 451.5px;
            height: 301px;
            margin-left: 0px;
            margin-top: 0px;
            transform: rotate(0rad) translateZ(0px);
            -webkit-transform: rotate(0rad) translateZ(0px);
          "
          title=""
      /></span>
    </p>
    <p class="c6 c4">
      <span class="c9 c8">Does Usain Bolt ever really touch the ground?</span>
    </p>
    <p class="c6 c4">
      <span class="c11 c12">Streaming Speedup: 35% faster</span>
    </p>
    <p class="c6 c4">
      <span class="c1"
        >This workflow is constant disk space and constant memory. This means
        you have a smaller digital footprint and you have your count matrices
        made before your friend has even downloaded their data.</span
      >
    </p>
    <h2 class="c19 c4" id="h.oveggaj71d5v">
      <span class="c11 c14">Comparison with Cell Ranger</span>
    </h2>
    <p class="c6 c4">
      <span
        >To understand the significance of this workflow I examined the disk
        usage of </span
      ><span class="c11">Cell Ranger</span><span>&nbsp;using the</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/pachterlab/MBGBLHGP_2019/tree/master/Supplementary_Figure_9&amp;sa=D&amp;source=editors&amp;ust=1671999727987410&amp;usg=AOvVaw0k2Vp5l8wo2PuYiUgxEihj"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/pachterlab/MBGBLHGP_2019/tree/master/Supplementary_Figure_9&amp;sa=D&amp;source=editors&amp;ust=1671999727987708&amp;usg=AOvVaw2E2G898n6z9NkiMZc4urhC"
          >Snakemake benchmark.</a
        ></span
      ><span class="c1"
        >&nbsp;Here are the results for 20 datasets ordered by increasing number
        of reads:</span
      >
    </p>
    <a id="t.95f8cfbd3266eaed54b15ffeb747a982d507a527"></a><a id="t.0"></a>
    <table class="c20">
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c17 c4"><span class="c11">Dataset</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c17 c4">
            <span class="c11">kallisto | bustools [Gb]</span>
          </p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c4 c17"><span class="c11">Cell Ranger [Gb]</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3 c4"><span class="c1">SRR8599150_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">0.003</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">17.90</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8611943_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">0.41</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">29.16</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR6998058_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">0.68</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">76.25</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">hgmm1k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1.26</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">113.15</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">pbmc1k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1.20</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">108.46</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">hgmm1k_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1.78</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">126.63</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">heart1k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1.80</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">136.19</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8206317_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">2.03</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">154.11</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">heart1k_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">2.05</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">129.82</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8524760_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1.96</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">158.23</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR7299563_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">2.01</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">181.70</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8513910_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">3.94</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">271.85</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR6956073_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">3.29</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">199.79</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8257100_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">5.32</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">330.09</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8327928_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">3.64</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">328.36</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">EMTAB7320_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">8.22</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">698.53</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">neuron10k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">6.99</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">577.88</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">SRR8639063_v2</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">10.56</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">617.51</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">pbmc10k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">11.28</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1,022.07</span></p>
        </td>
      </tr>
      <tr class="c13">
        <td class="c7" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">hgmm10k_v3</span></p>
        </td>
        <td class="c10" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">15.73</span></p>
        </td>
        <td class="c5" colspan="1" rowspan="1">
          <p class="c3"><span class="c1">1,310.56</span></p>
        </td>
      </tr>
    </table>
    <p class="c6 c4">
      <span
        >The number indicates the number of bytes written to disk. In the case
        of kallisto, it is the size of the BUS file and accessory files </span
      ><span class="c8">only, </span
      ><span>since no temporary files are written to disk. </span
      ><span class="c12 c11"
        >Cell Ranger may use well over a terabyte of disk space for
        pre-processing FASTQs.</span
      >
    </p>
    <h2 class="c19 c4" id="h.wu9j5s9sg385">
      <span class="c14 c11"
        >Stream arbitrarily large datasets from anywhere</span
      >
    </h2>
    <p class="c6 c4">
      <span
        >This workflow is easily extended to reads that are hosted on the ENA.
        An exercise for the reader: stream reads from</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.ebi.ac.uk/ena/data/view/SRX5245088&amp;sa=D&amp;source=editors&amp;ust=1671999728006627&amp;usg=AOvVaw2utV7-DM-zq_mG8lvEuTFB"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://www.ebi.ac.uk/ena/data/view/SRX5245088&amp;sa=D&amp;source=editors&amp;ust=1671999728006991&amp;usg=AOvVaw2u-0idkFQqR027BH8SJgMu"
          >this ENA dataset</a
        ></span
      ><span>&nbsp;to make a count matrix.</span
      ><span
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/sbooeshaghi/web_log/blob/master/2019_07_07-FASTerQ_to_Count_Matrices_for_Single_Cell_RNA_Seq/solution.md&amp;sa=D&amp;source=editors&amp;ust=1671999728007418&amp;usg=AOvVaw0rWVEXNUnoe4qmaVQtyEym"
          >&nbsp;</a
        ></span
      ><span class="c16"
        ><a
          class="c2"
          href="https://www.google.com/url?q=https://github.com/sbooeshaghi/web_log/blob/master/2019_07_07-FASTerQ_to_Count_Matrices_for_Single_Cell_RNA_Seq/solution.md&amp;sa=D&amp;source=editors&amp;ust=1671999728007805&amp;usg=AOvVaw0bWmJVFRhdwEkXPCEvnVJ-"
          >Here is the solution</a
        ></span
      ><span class="c1"
        >. Even though the dataset is composed of 505,958,821 reads you can
        produce count matrices on your laptop in a few hours!
      </span>
    </p>
    <p class="c3 c4 c15"><span class="c1"></span></p>
  </body>
</html>
